File: src/ast.ts
import type { Literal, TokenType } from "./lexer.ts";

export type NodeType =
  | "Program"
  | "Val"
  | "BinOp"
  | "UnaryOp"
  | "VarDecl"
  | "VarMod"
  | "Var"
  | "Out"
  | "In"
  | "Label"
  | "Goto"
  | "If"
  | "While"
  | "Del"
  | "Stmt"
  | "ExprStmt"
  | "FuncDecl"
  | "FuncCall"
  | "NativeFunc"
  | "Return"
  | "Exit";

export interface Node {
  type: NodeType;
}
export interface Program extends Node {
  type: "Program";
  body: Expression[];
}

export type Expression =
  | Val
  | BinOp
  | UnaryOp
  | VarDecl
  | VarMod
  | Var
  | Out
  | In
  | If
  | While
  | Label
  | Goto
  | Del
  | Stmt
  | FuncDecl
  | FuncCall
  | Return
  | Exit;

export interface Val extends Node {
  type: "Val";
  value: Literal;
}

export interface BinOp extends Node {
  type: "BinOp";
  op: TokenType;
  left: Expression;
  right: Expression;
}

export interface UnaryOp extends Node {
  type: "UnaryOp";
  op: TokenType;
  argument: Expression;
}

export interface VarDecl extends Node {
  type: "VarDecl";
  name: string;
  value: Expression;
}

export interface VarMod extends Node {
  type: "VarMod";
  name: string;
  value: Expression;
}

export interface Var extends Node {
  type: "Var";
  name: string;
}

export interface Out extends Node {
  type: "Out";
  value: Expression;
}

export interface In extends Node {
  type: "In";
  name: string;
}

export interface If extends Node {
  type: "If";
  cond: Expression;
  body: Expression;
  else?: Expression;
}

export interface While extends Node {
  type: "While";
  cond: Expression;
  body: Expression;
}

export interface Label extends Node {
  type: "Label";
  name: string;
}

export interface Goto extends Node {
  type: "Goto";
  name: string;
}

export interface Del extends Node {
  type: "Del";
  name: string;
}

export interface Exit extends Node {
  type: "Exit";
  status: Expression;
}

export interface Stmt extends Node {
  type: "Stmt";
  body: Expression[];
}

export interface FuncDecl extends Node {
  type: "FuncDecl";
  name: string;
  params: string[];
  body: Expression;
}

export interface FuncCall extends Node {
  type: "FuncCall";
  name: string;
  args: Expression[];
}

export interface NativeFunc extends Node {
  type: "NativeFunc";
  fn: (...args: Literal[]) => Literal | Promise<Literal>;
} // typescript functions

export interface Return extends Node {
  type: "Return";
  value: Expression;
}

File: src/ast.rs
use crate::token::{Literal, TokenType, VarType};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Program {
    pub body: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Stmt {
    pub body: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Expr {
    Stmt(Stmt),
    Val(Val),
    Var(Var),
    ArrayAccess(ArrayAccess),
    ArrayAssign(ArrayAssign),
    VarDecl(VarDecl),
    VarMod(VarMod),
    BinOp(BinOp),
    UnaryOp(UnaryOp),
    If(If),
    While(While),
    For(For),
    FuncDecl(FuncDecl),
    FuncCall(FuncCall),
    Return(Return),
    Label(Label),
    Goto(Goto),
    Extern(Extern),
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Val {
    pub value: Literal,
    pub typ: VarType,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Var {
    pub name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct VarDecl {
    pub name: String,
    pub value: Box<Expr>,
    pub typ: VarType,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct VarMod {
    pub name: String,
    pub value: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub operator: TokenType,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct UnaryOp {
    pub argument: Box<Expr>,
    pub operator: TokenType,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct If {
    pub condition: Box<Expr>,
    pub then: Box<Expr>,
    pub else_branch: Option<Box<Expr>>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct While {
    pub condition: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct For {
    pub init: String,
    pub iter: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct FuncDecl {
    pub name: String,
    pub params: Vec<String>,
    pub body: Box<Expr>,
    pub is_pub: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct FuncCall {
    pub name: String,
    pub args: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Return {
    pub value: Option<Box<Expr>>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Label {
    pub name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Goto {
    pub label: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct ArrayAccess {
    pub array: String,
    pub offset: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct ArrayAssign {
    pub array: String,
    pub offset: Box<Expr>,
    pub value: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Extern {
    pub func: String,
}
